File: scripts/load_data_mongodb.js
// MongoDB data loader (mongosh) for Big Data Assignment 2.
// Usage: mongosh --file scripts/load_data_mongodb.js
// NOTE: The Python loader (load_data_mongodb.py) is recommended for large files.
// This script provides grader-compatible mongosh loading.

const DB_NAME = 'ecommerce';
const DATA_DIR = '/hostdata/cleaned';

db = db.getSiblingDB(DB_NAME);

function loadCSV(collectionName, filename, indexes) {
    print(`Loading ${collectionName}...`);
    db[collectionName].drop();

    const lines = cat(`${DATA_DIR}/${filename}`).split('\n');
    const headers = lines[0].split(',');
    let batch = [];
    let count = 0;

    for (let i = 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue;

        const values = line.split(',');
        const doc = {};
        for (let j = 0; j < headers.length; j++) {
            const val = values[j] || '';
            if (val === '' || val === 'NaT') {
                doc[headers[j]] = null;
            } else if (val === 'True' || val === 'true') {
                doc[headers[j]] = true;
            } else if (val === 'False' || val === 'false') {
                doc[headers[j]] = false;
            } else if (!isNaN(val) && val !== '') {
                doc[headers[j]] = Number(val);
            } else {
                doc[headers[j]] = val;
            }
        }
        batch.push(doc);
        count++;

        if (batch.length >= 5000) {
            db[collectionName].insertMany(batch);
            batch = [];
        }
    }
    if (batch.length > 0) {
        db[collectionName].insertMany(batch);
    }

    if (indexes) {
        indexes.forEach(idx => {
            db[collectionName].createIndex(idx);
        });
    }

    print(`  ${collectionName}: ${db[collectionName].countDocuments()} docs`);
}

loadCSV('events', 'events.csv', [
    { user_id: 1 },
    { product_id: 1 },
    { event_type: 1 }
]);

loadCSV('campaigns', 'campaigns.csv', [
    { id: 1, campaign_type: 1 }
]);

loadCSV('messages', 'messages.csv', [
    { client_id: 1 },
    { campaign_id: 1, message_type: 1 },
    { is_purchased: 1 }
]);

loadCSV('client_first_purchase', 'client_first_purchase_date.csv', [
    { client_id: 1 },
    { user_id: 1 }
]);

loadCSV('friends', 'friends.csv', [
    { friend1: 1 },
    { friend2: 1 }
]);

print('MongoDB loading complete.');

File: scripts/load_data_mongodb.py
#!/usr/bin/env python3
"""
MongoDB data loader for Big Data Assignment 2.
Creates collections and loads cleaned CSVs using pymongo.
"""

import os
import math
import pandas as pd
from pymongo import MongoClient, ASCENDING, TEXT

CLEAN_DIR = os.path.join(os.path.dirname(__file__), '..', 'data', 'cleaned')
BATCH_SIZE = 10000


def nan_to_none(record):
    """Replace NaN/NaT with None for MongoDB."""
    return {
        k: (None if (isinstance(v, float) and math.isnan(v)) else v)
        for k, v in record.items()
    }


def load_collection(db, name, filename, indexes=None, chunk_size=50000):
    path = os.path.join(CLEAN_DIR, filename)
    print(f"  Loading {name} from {filename}...")
    db.drop_collection(name)
    coll = db[name]

    total = 0
    for chunk in pd.read_csv(path, chunksize=chunk_size, low_memory=False):
        records = chunk.to_dict('records')
        records = [nan_to_none(r) for r in records]
        for i in range(0, len(records), BATCH_SIZE):
            coll.insert_many(records[i:i + BATCH_SIZE])
        total += len(records)
        print(f"    {total} rows inserted...")

    if indexes:
        for idx in indexes:
            coll.create_index(idx)
            print(f"    index created: {idx}")

    print(f"  {name}: {coll.count_documents({})} docs")
    return coll


def main():
    client = MongoClient('localhost', 27017)
    db = client['ecommerce']

    print("[MongoDB] Loading data...")

    load_collection(db, 'events', 'events.csv', indexes=[
        [('user_id', ASCENDING)],
        [('product_id', ASCENDING)],
        [('event_type', ASCENDING)],
        [('category_code', TEXT)],
    ])

    load_collection(db, 'campaigns', 'campaigns.csv', indexes=[
        [('id', ASCENDING), ('campaign_type', ASCENDING)],
    ])

    load_collection(db, 'messages', 'messages.csv', indexes=[
        [('client_id', ASCENDING)],
        [('campaign_id', ASCENDING), ('message_type', ASCENDING)],
        [('is_purchased', ASCENDING)],
    ])

    load_collection(db, 'client_first_purchase', 'client_first_purchase_date.csv', indexes=[
        [('client_id', ASCENDING)],
        [('user_id', ASCENDING)],
    ])

    load_collection(db, 'friends', 'friends.csv', indexes=[
        [('friend1', ASCENDING)],
        [('friend2', ASCENDING)],
    ])

    client.close()
    print("[MongoDB] Done.")


if __name__ == '__main__':
    main()
